README.md
# The great API API
## v1.1.0

The great API API is an API that allows users to find, test and share API's. 
With The great API, you can find API's for your project in seconds with all the usefull information and endpoints you need for development. 

---
## Why make an API API?
This project is a school project first, but a great tool nonetheless. Sometimes we look for API's, but not finding what we are looking for can be frustrating. 

---
## Getting started
1. Clone this repository.
2. Change directory to the api folder and install dependancies 
    ``` 
    cd api 
    npm install
    ```
3. Start up the server
    ```
    npm start
    ```
4. (TODO: Add info about postgreSQL server)
---
## How it works
The Great API has multiple endpoints to use. 

- `GET /apis`
    - Retrieves all the entries from the database
    - Has pagination
        - example request:  `GET /apis?page=1&limit=5`

- `GET /apis/:uuid`
    - Retrieves the entry where the uuid matches the provided uuid

- `POST /apis`
  - Inserts a new entry in the database
  - Body: 
    ``` 
        api_name: String
        api_url: String
        allowed_endpoints: String[]
        description: String
    ```



- `PATCH /apis/:uuid`
  - Edit **one** or **multiple** properties from a previous entry that matches the provided uuid
  - You don't have to provide all the other properties in the body, only the ones you wish to update
  - Example: 
    ```
    patchBody = {
        api_name: "New api name",
        description: "this description has been updated"
    }

    ```

- `DELETE /apis/:uuid`
  - Removes the entry from the database that matches the provided uuid
---
## Status & Roadmap

In development.

---
## Author 

Julien Menten

---
## Having trouble? 

Check out the contributing guidelines for reporting issues or setting up the development environment. 

---


CHANGELOG.md
# Changelog 

## V1.1.0
- Added Pagination to `GET /apis`
- Solved helpers tests
## V1.0.0
- Initialized project

- Added all endpoints 
    - `GET /apis`
    - `GET /apis/:uuid`
    - `POST /apis/`
    - `PATCH /apis/:uuid`
    - `DELETE /apis/:uuid`

- Added fetch API function
- Wrote integration test for Database from `POST` to `DELETE`
- Added small easter egg 
    - `GET /teapot`

api/src/__tests__/server.test.js
const e = require('express');
const Server  = require('http');
const supertest = require('supertest');
const app = require('../server');

const request = supertest(app)

 // Variables
let UUID = "";
const TEST_API = {
    api_url: "https://jsonplaceholder.typicode.com/todos/1",
    api_name: "JSON placeholder test API",
    allowed_endpoints: ["GET", "POST", "PUT", "PATCH", "DELETE"],
    description: "Free to use fake online REST API for testing and prototyping."
}

describe(' GET /api', ()=> {
    test('responds with 200', async done => {
         const response = await request.get('/apis')
         expect(response.status).toBe(200)
         done()
    })
})

describe('Create and POST a new API in the database, then remove that entry', () => {

    test('POST a new entry to the DB', async done => {
        const response = await request.post('/apis').send(TEST_API)
        UUID = response.body.id
       
        expect(response.status).toBe(200)
        done()
    })

    test('GET the new entry from the DB', async done => {
        const response = await request.get(`/apis/${UUID}`)
        expect(response.status).toBe(200)
        done()
    })

    test('UPDATE the name of the entry', async done => {
        const response = await request.patch(`/apis/${UUID}`)
            .send({
                api_name: "Test PATCH"
            })
        expect(response.status).toBe(200)
        done()
    })

   
})


describe('DELETE /apis/:id', () => {
    test('DELETE the entry', async done => {
        const response = await request.delete(`/apis/${UUID}`)
        expect(response.status).toBe(200)
        done()
    })
    
});


describe('GET /apis/:id', () => {
    it('should return "API not found" when an incorrect id has been provided', async done => {

        const fakeId = "b596d3f5-c83e-4402-a54a-756fc7f7efe1"

        const response = await request.get(`/apis/${fakeId}`)
        expect(response.status).not.toBe(200)
        expect(response.text).toBe("API not found");
        done()
    });
})

api/src/__tests__/helpers.test.js
const apiHelpers = require('../helpers/apiHelpers');


describe('Test if we can fetch data from a random API online', () => {
    it('should return a status of 200 and a stringified content', async done => {
        const TEST_URL = "https://api.chucknorris.io/jokes/random";
        let api = await apiHelpers.fetchNewAPIData(TEST_URL);
      
        expect(api)
        .toStrictEqual(
            expect.objectContaining({
                properties : expect.any(String),
                status: expect.any(Number)
            })
        )
        done()
    })
});

describe("Test the ability to analyze a JSON file and it's properties and their types" , () => { 
    const testUser = {
        name: "JohnDoe",
        age: 26,
        hobbies: ['football', 'games']
    }
    test("Read the properties of an object", async done => {
        expect(apiHelpers.formatProperties(testUser)).toEqual(
            expect.arrayContaining([{attribute_name:'name', attribute_type: 'String'}, {attribute_name:'age', attribute_type: 'Number'}, {attribute_name:'hobbies', attribute_type: "Object"}])
        )
        expect(apiHelpers.formatProperties(testUser)).not.toEqual(
            expect.arrayContaining(['username', 'birthday', 'hobbies'])
        )
        done()
    })

    test("Analyzes the properties of an object and returns the type of the property", async done => {
        expect(apiHelpers.formatProperties(testUser)).toEqual(
            expect.arrayContaining([{attribute_name:'name', attribute_type: 'String'}, {attribute_name:'age', attribute_type: 'Number'}, {attribute_name:'hobbies', attribute_type: 'Object'}])
        )
        done()
    })

    test("Read an API online and analyze the properties and return their types and names", async done => {


        done()
    })
})

describe("Test if the requested API returns an Array of results, or only a simple JSON" , () => { 
    const TestArray = [1, 2, 3]
    const TestJSON = {
        "username": "John Doe",
        "age": 24
    }
    it("should return 'false' in case it is not an array", (done) => {
        expect(apiHelpers.detectIsArray(TestJSON)).not.toBeTruthy()
        expect(apiHelpers.detectIsArray(TestJSON)).toBeFalsy()
        done()
    })
    it("Should return 'true' when the inut is an array", (done) => {
        expect(apiHelpers.detectIsArray(TestArray)).toBeTruthy()
        done()
    })
})

api/src/__tests__/integration/server.test.js
const e = require('express');
const Server  = require('http');
const supertest = require('supertest');
const app = require('../../server');

const request = supertest(app)

 // Variables
let UUID = "";
const TEST_API = {
    api_url: "https://jsonplaceholder.typicode.com/todos/1",
    api_name: "JSON placeholder test API",
    allowed_endpoints: ["GET", "POST", "PUT", "PATCH", "DELETE"],
    description: "Free to use fake online REST API for testing and prototyping."
}

describe(' GET /api', ()=> {
    test('responds with 200', async done => {
         const response = await request.get('/apis')
         expect(response.status).toBe(200)
         done()
    })
})

describe('Create and POST a new API in the database, then updates the name', () => {

    test('POST a new entry to the DB', async done => {
        const response = await request.post('/apis').send(TEST_API)
        UUID = response.body.id
       
        expect(response.status).toBe(200)
        done()
    })

    test('GET the new entry from the DB', async done => {
        const response = await request.get(`/apis/${UUID}`)
        expect(response.status).toBe(200)
        done()
    })

    test('UPDATE the name of the entry', async done => {
        const response = await request.patch(`/apis/${UUID}`)
            .send({
                api_name: "Test PATCH"
            })
        expect(response.status).toBe(200)
        done()
    })

   
})


describe('DELETE /apis/:id', () => {
    test('DELETE the entry', async done => {
        const response = await request.delete(`/apis/${UUID}`)
        expect(response.status).toBe(200)
        done()
    })
    
});


describe('GET /apis/:id', () => {
    it('should return "API not found" when an incorrect id has been provided', async done => {

        const fakeId = "b596d3f5-c83e-4402-a54a-756fc7f7efe1"

        const response = await request.get(`/apis/${fakeId}`)
        expect(response.status).not.toBe(200)
        expect(response.text).toBe("API not found");
        done()
    });
})

api/src/helpers/apiHelpers.js
const fetch = require('node-fetch');
   /*
 =================================================================
        FUNCTIONS MANIPULATING DATA
 =================================================================
*/
const ApiHelpers = {
    /* 
    Fetches the API provided at POST and uses that data to find the 
        Input: 
            URL: string

    This either returns the whole JSON object, or the first entry from the fetched data   
    */
    async fetchNewAPIData(URL) {
        let result 
        try{
            await fetch(URL)
            .then(async data => {
                let content = await data.json()
                
                if(this.detectIsArray(content)){
                    let firstOfArray = content[0]
                    console.log(firstOfArray)
                    return{
                           properties: firstOfArray,
                           status: data.status
                        }
                }else {
                    return {    
                            properties: content,
                            status: data.status
                        }
                }
               
            })
            .then(res => {
                let contentString = JSON.stringify(res.properties)
                return result = {
                    properties: contentString,
                    status: res.status
                }
            })
          
        }catch(e){
            console.error(e)
        }
        return result
    },
    /* 
    Analyzes an api and its attributes. Then creates a list from the attributes it found and their type. 
    This list is then returned as an array containing objects. 
        Input: 
            data: 1 result object from the fetched API 

        Output:   
            properties: {
                attribute_name: String
                attribute_type: String
            }[]
        
    */
    formatProperties(data){
        if(!undefined) {
            let properties = Object.keys(data);
            let test = Object.values(data)
            let formattedProperties = [];

            properties.forEach(prop => {
                formattedProperties.push({
                    attribute_name: prop,
                    attribute_type: this.getPropertyType(test[properties.indexOf(prop)])
                })
            })
            return formattedProperties;
        }else {
            return "error" 
        }
        
    },

    /*
    This function analyzes 1 property and returns it's type. 
        Input: 
            prop: any
        Output: 
            propType: String

        TODO: Add more possible outputs
    */
    getPropertyType(prop) {

        let propType = null
        switch (typeof prop) {
            case 'string':
                propType = "String"
                break;
            case 'number':
                propType = "Number"
                break;
            case 'boolean':
                propType = "Array"
                break;
            case 'object':
                propType = "Object"
                break;
            default:
                propType = "Undefined"
                break;
        }
        return propType;
    },

    /*
    This function checks if the input is either an array or an object of any kind. Returns boolean.
        Input:
            data: object or array
        Output: 
            boolean
    */
    detectIsArray(data) {
        if(Array.isArray(data)) {
            return true
        }
        else {
            return false
        }
    },

    /*
    paginate function takes the data model and returns a paginated version with the set limit that is passed through the url of the GET request.
    Used as middleware for the GET /apis request
    */
    paginate(data, page, limit) {
        let currentPage = parseInt(page);
        let resultLimit = parseInt(limit);
        let start = (currentPage - 1) * resultLimit;
        let end = currentPage * resultLimit;

        let result = {};

        if (end < data.length) {
            result.nextPage = {
                page: currentPage + 1,
                limit: resultLimit
            }
        }
        if (start > 0) {
            result.previousPage = {
                page: currentPage - 1,
                limit: resultLimit
            }
        }
        
        result.results = data.slice(start, end);
        paginatedResult = result;

        return paginatedResult
    }
}

module.exports = ApiHelpers;




api/src/helpers/DatabaseHelper.js

/*
 =================================================================
        FUNCTIONS MANIPULATING THE DATABASE
 =================================================================
*/
const pg = require('knex')({
    client: 'pg',
    version: '9.6',      
    connection: 'postgres://julien:admin12345@localhost:5432/thegreatapi',
    searchPath: ['knex', 'public'],
  });

const DatabaseHelper = {
/*
    Checks the database and it's tables. In case an 'api' table does not exist, it creates a new table that follows the knex schema. 
    This knex schema contains; 
        - id (increment): int
        - uuid: uuid
        - api_name: String
        - api_url: String
        - description: String
        - properties: String[]
        - endpoints: json
        - timestamps: Timestamps
*/
    async initialiseTables() {
        await pg.schema.hasTable('api').then(async (exists) => {
          if (!exists) {
            await pg.schema
              .createTable('api', (table) => {
                table.increments();
                table.uuid('uuid');
                table.string('api_name');
                table.string('api_url');
                table.uuid('category');
                table.string('description');
                table.json('properties');
                table.json('endpoints');
                table.timestamps(true, true);
              }).then( (res) => {
                    console.log(res)
                    console.log('created table: APIS');
              });
      
          } else {
              console.log('API table already exists')
          }
        });

        await pg.schema.hasTable('categories').then(async (exists) => {
            if(!exists) { 
                await pg.schema
                .createTableIfNotExists('categories', (table) => {
                    table.increments();
                    table.uuid('uuid');
                    table.string('category_name');
                    table.timestamps(true, true);
                })
            }
        })
      },


/* 
    ResetTables function resets the table in case a column has been changed, added or removed. 
    ! ONLY USE WHEN A RESET OF THE TABLE IS NEEDED
*/
    async resetTables(){
        await pg.schema.dropTableIfExists('api');
    },

    /*
    InsertAPI() function accepts an object with data comming from the POST request and inserts the new API record to the database.
    This function also searches for the properties from the API makes a readable JSON format to put in the database. 

    TODO: READ THE API AND FORMAT THE PROPERTIES IN A JSON FORMAT THAT IS READABLE FOR FRONT END USE.
*/ 
    async insertAPI(data){
        await pg('api').insert({
            uuid: data.uuid,
            api_name: data.api_name,
            api_url: data.api_url,
            properties: {"properties":data.properties},
            endpoints: {"endpoints":data.allowed_endpoints},
            description: data.description
        }).returning('*').then(newData => {
            console.log(`New API added! Welcome ${newData[0].api_name}`)
        })
    },
    
}

module.exports = DatabaseHelper, pg;

api/src/server.js
const express = require('express');
const http = require('http');
const bodyparser = require('body-parser')
const app = express();
const { json } = require('body-parser');
const {v4: uuidv4} = require('uuid')
const DatabaseHelper = require('./helpers/DatabaseHelper');
const ApiHelpers = require('./helpers/apiHelpers');
const _ = require('lodash')
/*
 */
const pg = require('knex')({
    client: 'pg',
    version: '9.6',      
    connection: 'postgres://julien:admin12345@localhost:5432/thegreatapi',
    searchPath: ['knex', 'public'],
  });



// Middelware
app.use(bodyparser.json())
app.use(bodyparser.urlencoded({extented: true}))

/*
 =================================================================
            API ENDPOINTS
 =================================================================
*/
/*
GET /apis
  Gets all the records from the 'api' Database
  Uses paginate() function from ApiHelpers to set a limit to the result and paginate the results to avoid clutter
*/
app.get('/apis', async (req, res) => {
    let apiList = []
    try {
            await pg('api').select('*').then( apis => {
            apis.forEach( api => {
                apiList.push(api)
            })
        })
        if(!(_.isEmpty(req.query))){
            let paginatedResult = ApiHelpers.paginate(apiList, req.query.page, req.query.limit)
            res.status(200).send(paginatedResult)
        }
        else {
            res.status(200).send(apiList)
        }
    }
    catch(e) {
        res.status(400).send(e)
    }
});


/*
GET /apis/:uuid
    Finds the API in the databse that matches the provided UUID
*/
app.get('/apis/:id', async (req, res) => {
    console.log
    const uuid = req.params.id;

    const result = await pg('api')
        .select('*')
        .where({uuid: uuid})

    if(result.length >= 1) {
        res.status(200).send({
            result: result
        });
    }else {
        res.status(400).send("API not found")
    }
    
});


/* 
POST /apis
    Adds a new API to the database.
    The body of the request must contain:
        API-name: String
        API-link: String
        Allowed-endpoints: String[]
        Description: String
    
    TODO: Add security in case the request does not have the required inputs
*/
app.post('/apis' , async (req, res) => {

    if(req.body != undefined) {
        let api_properties = await ApiHelpers.fetchNewAPIData(req.body.api_url);
        let formattedProperties = ApiHelpers.formatProperties(JSON.parse(api_properties.properties));
        const newAPI = {
            uuid: uuidv4(),
            api_name: req.body.api_name,
            api_url: req.body.api_url,
            properties: formattedProperties,
            allowed_endpoints: req.body.allowed_endpoints,
            description: req.body.description
        }
        
        DatabaseHelper.insertAPI(newAPI);
        res.status(200).send({
            "status": 200,
            "message": "Record added!",
            "id": newAPI.uuid,
            "api": newAPI
        })
    } else {
       res.status(400).send("Invalid body received")
    }
   
})

/* 
DELETE /apis/:uuid
    Deletes a record from the database that matches the provided UUID
*/
app.delete('/apis/:uuid', async (req, res) => {
    const uuid = req.params.uuid
    try {
        await pg('api').where({uuid: uuid}).del().returning('*').then(data => {
            console.log(`API deleted! Goodbye`)
            res.status(200).send()
        })
        
    }catch(e){
        console.log(e)
        res.status(400).send("An unknown error has occured")
    }
    
})

/* 
PATCH /apis/:uuid
   Updates the content of a specific API that matches the provided UUID. 
   The body can be any attributes from the original object: 
        API-name: String
        API-link: String
        Allowed-endpoints: String[]
        Description: String

        TODO: Secure the endpoint so that you get error messages 
*/

app.patch('/apis/:uuid', async (req, res) => {
    const uuid = req.params.uuid;
    const newBody = req.body
    try{    
        await pg('api').where({uuid: uuid}).update(newBody).returning('*').then(data => {
            res.status(200).send()
        })
    }  catch(e) {
        res.status(400).send()
    }
})


app.get('/teapot', (req, res) => {
    res.status(418).send()
})


/*
    Table functions
*/
DatabaseHelper.initialiseTables()

module.exports = app;